### ISSUE TITLE
problems with as.difftime

### ISSUE THREAD

**Issue by chiefmurph** [chiefmurph/mondate#2]
## email 11/2/2015 10:19pm from Wu Junzhou:

Dear Murphy,

I encountered a problem when trying to calculate years between two dates.Here is the code:

> require(mondate)
> date1 = mondate("2015-12-31")
> date2 = mondate("2013-12-01")
> as.difftime(date1-date2)
> Time difference of 24.96774 months
> as.difftime(date1-date2,units = "years")
> Error in as.difftime(date1 - date2, units = "years") : 
>   'tim' must be numeric for units = 'years'
> 
> sessionInfo()
> R version 3.2.2 (2015-08-14)
> Platform: x86_64-w64-mingw32/x64 (64-bit)
> Running under: Windows 7 x64 (build 7601) Service Pack 1

locale:
[1] LC_COLLATE=Chinese (Simplified)_People's Republic of China.936 
[2] LC_CTYPE=Chinese (Simplified)_People's Republic of China.936  
[3] LC_MONETARY=Chinese (Simplified)_People's Republic of China.936
[4] LC_NUMERIC=C  
[5] LC_TIME=Chinese (Simplified)_People's Republic of China.936    

attached base packages:
[1] stats     graphics  grDevices utils     datasets  methods   base     

other attached packages:
[1] mondate_0.10.01.02

loaded via a namespace (and not attached):
[1] tools_3.2.2

I have found no solution with  this problem in the internet.

I hope to look forward for your support

sincerely,

Wu Junzhou

**Comment by chiefmurph** [comment]
## Replied Tue, Nov 3, 2015 at 4:39 AM

Dear Mr. Wu:

Thank you for your inquiry. The function is performing as documented. Here is the explanation.

In your first call
as.difftime(date1-date2)
there are two things going on:
(1) units = "auto" by default, and
(2) date1-date2 is not numeric but is a difftime object (see class(date1-date2)).
Because of (1), the 'base' version of as.difftime is called. Because of (2) the difftime object itself (date1-date2) is simply returned. Therefore, no error. Type base::as.difftime at the console to see the code, which is only a few lines.

In your second call
as.difftime(tim = date1-date2,units = "years")
units = "years" is specified, so the first argument, tim, must be numeric (per the documentation). But it is not numeric, it is a difftime object.

I am curious. Why do you calculate as.difftime(date1-date2,units = "years")?

Sincerely,
Dan Murphy

**Comment by chiefmurph** [comment]
## Replied Nov 3, '15, 6:12pm

Dear  Murphy,

Thank you very much for your reply.  I understand the function of mondate::as.difftime, which is not suit for time interval calculator.

So if I need to calculate the time interval by months between  two dates, just simply call

> date1 = mondate("2015-12-31")
> date2 = mondate("2013-12-01")
> (time_interval_by_months = date1 - date2)
> Time difference of 24.64516 months

However, my purpose is to calculate the time interval by years. my solution is:

> as.difftime(as.numeric(time_interval_by_months/12), , "years")
> Time difference of 2.053763 years

is there any better solution to calculate the time interval by years.?

Sincerely,
Wu Junzhou

**Comment by chiefmurph** [comment]
## Replied Nov 3, 2015, 6:42pm

You have uncovered a bug in my package. I will give you two solutions. The second does not work as intended due to a bug. Then I will give you a workaround

> # 1
> 
> date1 = mondate("2015-12-31")
> date2 = mondate("2013-12-01")
> (time_interval_by_months = date1 - date2)
> Time difference of 24.96774 months
> (time_interval_by_years = time_interval_by_months / 12)
> Time difference of 2.080645 months

The magnitude is correct, but the units still show as "months". Not a satisfactory solution. This next solution was supposed to give the correct answer:

> # 2
> 
> date1 = mondate("2015-12-31", timeunits = "years")
> date2 = mondate("2013-12-01", timeunits = "years")
> (time_interval_by_years = date1 - date2)
> Time difference of 24.96774 years

The units is correct, but the magnitude is incorrect. Here is the workaround:

> # workaround
> 
> date1 = mondate("2015-12-31")
> date2 = mondate("2013-12-01")
> (time_interval_by_years = as.difftime(as.numeric(date1 - date2) / 12, units = "years"))
> Time difference of 2.080645 years

I am in the middle of a rewrite of another section of the package and will fix this bug at the same time. Thank you for bringing the issue to my attention.

Sincerely,
Dan Murphy

**Comment by chiefmurph** [comment]
Dear Wu Junzhou:

It has been many years since you brought up this issue. As I look back, my reply recommending use of as.difftime was incorrect. The Github version of the package, when built on your PC, will generate your desired output:
> date1 = mondate("2015-12-31", timeunits = "years")
> date2 = mondate("2013-12-01", timeunits = "years")
> date1 - date2
Time difference of 2.080645 years

The CRAN version incorrectly gives
Time difference of 24.96774 years

I need to correct the CRAN version. That means I need to revert the mondate code back to the CRAN version 0.10.01.02 published 2013-07-21. Unfortunately, the Github version has strayed a long ways away from the CRAN version, including directions I may no longer support. Therefore, some functionality on GitHub may be lost. Most of the commit history since July 2013 has to do with creating vignettes. But some functionality like "cut" and "cutmondate" will be temporarily lost.

After fixing the bug you discovered, I intend to re-commit the cut functions and other useful functionality. I am not sure what vignette(s) will be retained. Hopefully any bumps in the road will not be severe.

Bottom line, please fork the Github repository if you wish to keep any post-v.0.10.01.02 functionality you find useful. 

Please let me know if you have any concerns. I will look for your reply over the next two weeks.

Thank you for your interest in mondate!

Sincerely,
Dan Murphy
